decoder/correlator.go

```diff
@@ -12,140 +12,309 @@ import (
+	// rtcp-endpoints is the average number of used rtcp endpoints per call.
+	// entry-size is the average size of one endpoint entry, including textual ip length, textual port length, ssrc-length, call-id length and two separators.
+	// Some guesses: concurrent-calls=10000, number-of-rtcp-endpoints=3, entry-size=1000.
 	rtcpCache = freecache.NewCache(30 * 1024 * 1024) // 30 MB
+	// sdpCacheTime is the maximum time between seeing sdp and seeing the first packets for all associated rtcp streams.
+	sdpCacheTime = 60 * 20 // 20 minutes in seconds.
+	// rtcpCacheTime is the maximum time a rtcp stream may be associated to a call (maximum allowed call time).
+	rtcpCacheTime = 60 * 60 * 60 // 60 hours in seconds.
 )
 
-// cacheSDPIPPort will extract the source IP, source Port from SDP body and CallID from SIP header.
-// It will do this only for SIP messages which have the strings "c=IN IP4 " and "m=audio " in the SDP body.
-// If there is one rtcp attribute in the SDP body it will use it as RTCP port. Otherwise it will add 1 to
-// the RTP source port. These data will be used for the SDPCache as key:value pairs.
+// addSDPCacheEntry will add an entry to SDPCache with rtcpIP+rtcpPort as key and callID as value.
+// Key parts will be separated by a single space.
+func addSDPCacheEntry(rtcpIP []byte, rtcpPort []byte, callID []byte) {
+	key := bytes.Join([][]byte{rtcpIP, rtcpPort}, []byte(" "))
+	logp.Debug("sdp", "Add to sdpCache key=%q, value=%q", key, callID)
+	sdpCache.Set(key, callID, sdpCacheTime)
+}
+
+// cacheSDPIPPort will extract the Call-ID and all RTCP ip and port combinations will add them to the sdpCache,
+// with ip+port as key and Call-ID as value.
+//
+// It will only process payload that has SDP content or multipart content that contains SDP.
+// It will only process audio media.
+// There must be a Call-ID in the SIP-Headers.
+// It will use ips from source, c lines and a=rtcp lines.
+// It will use rtcp ports from m lines (rtp port + 1) or a=rtcp lines.
+// It will only use the first address from multi address notation.
+// It will only use the first port from multi port notation.
+// The function makes some assumptions about the well-formedness of the SDP for faster parsing.
+// Key parts will be separated by a single space.
 func cacheSDPIPPort(srcIP net.IP, srcPort uint16, dstIP net.IP, dstPort uint16, payload []byte) {
-	if posSDPIP := bytes.Index(payload, cLine); posSDPIP > 0 {
-		if posSDPPort := bytes.Index(payload, mLine); posSDPPort > 0 {
-			ipPort.Reset()
-			restIP := payload[posSDPIP:]
-			// Minimum IPv4 length of "c=IN IP4 1.1.1.1" = 16
-			if posRestIP := bytes.Index(restIP, []byte("\r\n")); posRestIP >= 16 {
-				ipPort.Write(restIP[len(cLine)+2 : posRestIP])
-			} else {
-				logp.Debug("sdp", "No end or fishy SDP IP in '%s'", restIP)
-				return
+	// TODO: improve multipart handling.
+
+	srcIPb := []byte(srcIP.String()) // source ip as text as bytes.
```

Comment:
You shouldn't need this as net.IP type is a []byte so just pass srcIP

Reply:
Using the ip bytes directly will use the raw binary representation, not the textual representation. Mixing binary and textual representation does not work when comparing or doing key lookup. As sdp and currently the cache use the textual representation some conversion must be done.
The binary representation for ip and port could be used everywhere, which would reduce the conversions needed when processing rtcp in trade-off for increased conversion when processing the sdp.
Using binary would make the current debug output of the keys a little less useful, but the debug output could be adjusted accordingly.

Reply:
Yip I know. We could change the srcKey inside correlateRTCP func for the sdpcache to use bytes representation. Overall this should reduce at least 1 allocation. The log output can be easily adjusted.

Besides that I think we can merge and I will see if I find some time to add some tests for cacheSDPIPPort and correlateRTCP.

Reply:
More tests would be nice. Doing some test captures does not cover all possible cases and I do not want to break something.
It would also be helpful if the test data would not be completely be a hex list, as a human can not see what sip is tested.

Thanks for the fast/helpful responses. I currently have only limited time for this project, but needed a better working rtcp capturing. Having this in the official version helps me and hopefully other people too.